sdk/helper/random/parser.go

```diff
@@ -0,0 +1,250 @@
+	if err != nil {
+		return strs, fmt.Errorf("unable to parse rules: %w", err)
+	}
+
+	// Add any charsets found in rules to the overall charset & deduplicate
+	cs := append(strs.Charset, getChars(rules)...)
+	cs = deduplicateRunes(cs)
+
+	strs = StringGenerator{
+		Length:  strs.Length,
+		Charset: cs,
+		Rules:   rules,
+	}
+
+	err = validate(strs)
+	if err != nil {
+		return strs, err
+	}
+
+	return strs, nil
+}
+
+func (p Parser) parseRules(rawRules []map[string]interface{}) (rules []Rule, err error) {
+	for _, rawRule := range rawRules {
+		info, err := getRuleInfo(rawRule)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get rule info: %w", err)
+		}
+
+		// Map names like "lower-alpha" to lowercase alphabetical characters
+		applyShortcuts(info.data)
+
+		rule, err := p.RuleRegistry.parseRule(info.ruleType, info.data)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse rule %s: %w", info.ruleType, err)
+		}
+		rules = append(rules, rule)
+	}
+
+	return rules, nil
+}
+
+func validate(strs StringGenerator) (err error) {
+	merr := &multierror.Error{}
+	if strs.Length < 1 {
+		merr = multierror.Append(merr, fmt.Errorf("length must be >= 1"))
+	}
+	if len(strs.Charset) == 0 {
+		merr = multierror.Append(merr, fmt.Errorf("no charset specified"))
+	}
+
+	for _, r := range strs.Charset {
+		if !unicode.IsPrint(r) {
+			merr = multierror.Append(merr, fmt.Errorf("non-printable character in charset"))
+			break
+		}
+	}
+
+	return merr.ErrorOrNil()
+}
+
+func getChars(rules []Rule) (chars []rune) {
+	type charsetProvider interface {
+		Chars() []rune
+	}
+
+	for _, rule := range rules {
+		cv, ok := rule.(charsetProvider)
+		if !ok {
+			continue
+		}
+		chars = append(chars, cv.Chars()...)
+	}
+	return chars
+}
+
+// getMapSlice from the provided map. This will retrieve and type-assert a []map[string]interface{} from the map
+// This will not error if the key does not exist
+// This will return an error if the value at the provided key is not of type []map[string]interface{}
+func getMapSlice(m map[string]interface{}, key string) (mapSlice []map[string]interface{}, err error) {
+	rawSlice, exists := m[key]
+	if !exists {
+		return nil, nil
+	}
+
+	slice, ok := rawSlice.([]map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("key %s is not a []map[string]interface{}", key)
+	}
+
+	return slice, nil
+}
+
+type ruleInfo struct {
+	ruleType string
+	data     map[string]interface{}
+}
+
+// getRuleInfo splits the provided HCL-decoded rule into its rule type along with the data associated with it
+func getRuleInfo(rule map[string]interface{}) (data ruleInfo, err error) {
+	// There should only be one key, but it's a dynamic key yay!
+	for key := range rule {
+		slice, err := getMapSlice(rule, key)
+		if err != nil {
+			return data, fmt.Errorf("unable to get rule data: %w", err)
+		}
+		data = ruleInfo{
+			ruleType: key,
+			data:     slice[0],
+		}
+		return data, nil
+	}
+	return data, fmt.Errorf("rule is empty")
+}
+
+var (
```

Comment:
It might make sense to make these charsets available in the SDK so we can source them in other places such as https://github.com/hashicorp/vault/blob/cb178b7e4fd54369331000141fa7820cd94e6f29/sdk/helper/base62/base62.go#L12

Reply:
Good idea. Done.